### PR TITLE
Document any implicit filtering

### PR DESCRIPTION
**Pull request by ddufourboivin-zalando**
Add new rule 226 "Document any implicit filtering" (#428).

### CODE REVIEW

File: chapters/http-requests.adoc [hunk 1]
```diff
@@ -252,3 +252,67 @@ parameters as described below.
 When choosing the collection format, take into account the tool support,
 the escaping of special characters and the maximal URL length.
 
+[#226]
+== {MUST} Document any implicit filtering
```
Comment (tkrop) [inline chapters/http-requests.adoc]: As far as I know we prefer titles with capital letters.

I would remove `any`.

File: chapters/http-requests.adoc [hunk 2]
```diff
@@ -252,3 +252,67 @@ parameters as described below.
...
+
+Implicit filtering should be avoided as much as possible. However, in some
+cases (usually for legal reasons), some endpoints should not be listing all
+resources they have, but only those on which the current user has access to
```
Comment (tkrop) [inline chapters/http-requests.adoc]: `user` -> `consumer`

File: chapters/http-requests.adoc [hunk 3]
```diff
@@ -252,3 +252,67 @@ parameters as described below.
...
+(or some other context specific condition).
```
Comment (tkrop) [inline chapters/http-requests.adoc]: Are there other conditions?

File: chapters/http-requests.adoc [hunk 4]
```diff
@@ -252,3 +252,67 @@ parameters as described below.
...
+and transparently in the API Specification (in its description).
+
+Example:
+
+If an employee of the company _Foo_ accesses one of our business-to-business
+service and performs a `GET /business-partners`, it must, for legal reasons, not display
+any other business partner that is not owned or contractually managed by her/his company.
+It should never see that we are doing business also with company _Bar_.
+
+[source]
```
Comment (tkrop) [inline chapters/http-requests.adoc]: In my humble opinion the before sentence is example enough. Anyhow, I would favor a shorter examples than full blown `curl` output.

File: chapters/http-requests.adoc (hunk 3 above)
Comment (ddufourboivin-zalando) [reply chapters/http-requests.adoc]: There could be, so I would not limit it exclusively to this (legal reasons).

File: chapters/http-requests.adoc [hunk 5]
```diff
@@ -252,3 +252,61 @@ parameters as described below.
...
+        { "name": "Bar pour Elle" }
+    ]
+}
+----
+
+The API Specification should then specify something like this:
+
+[source,yaml]
+----
+  /busines-partner:
```
Comment (dehora) [inline chapters/http-requests.adoc]: `busines-` -> `business`

File: chapters/http-requests.adoc [hunk 6]
```diff
@@ -252,3 +252,61 @@ parameters as described below.
 When choosing the collection format, take into account the tool support,
 the escaping of special characters and the maximal URL length.
 
+[#226]
+== {MUST} Document Implicit Filtering
+
+Implicit filtering should be avoided as much as possible. However, in some
+cases (usually for legal reasons), some endpoints should not be listing all
+resources they have, but only those on which the current consumer has access
+to (or some other context specific condition).
```
Comment (dehora) [inline chapters/http-requests.adoc]:
```patch
+ Implicit filtering should be avoided as much as possible. However, in some
+ cases (usually for legal reasons), some endpoints should not be listing all
```

I don't think the should qualifier is needed. This implicit filtering is standard behaviour for product APIs that support multiple tenants, especially SaaS products (eg Stripe/Intercom/GitHub/Twilio all do this) that document a single set of resources, bearing in mind that API as a product is a top level ambition. The leading sentence could be

```
Sometimes certain collection resources or queries will not list all the 
possible elements they have, but only those for which the current client 
is authorized to access.
```

File: chapters/http-requests.adoc [hunk 7]
```diff
@@ -252,3 +252,61 @@ parameters as described below.
...
+
+Implicit filtering could be done on:
+
+* the list of resources being return on a parent `GET` request
+* the fields returned for the resource's detail
+
+In such cases, the implicit filtering must absolutely be documented clearly
+and transparently in the API Specification (in its description).
```
Comment (dehora) [inline chapters/http-requests.adoc]: I would expect some guidance on caching in these cases. It's generally not safe for middleboxes and proxies to blindly cache results against a common resource URI without something like a vary header on the conditional element of the request (eg the Authorization or Cookie header). That gets into more sophisticated and more complicated caching techniques which might be best avoided.

Comment (whiskeysierra) [reply chapters/http-requests.adoc]: > If an API is intended to support caching, it must take care to specify this ability by defining the caching boundaries i.e. time-to-live and cache constraints, by providing the Cache-Control and Vary headers
> [..]
> Caching has to take many aspects into account, e.g. general cacheability of response information, our guideline to protect endpoints using SSL and OAuth authorization
>
> https://opensource.zalando.com/restful-api-guidelines/#_must_document_caching_if_supported

I'd just link there. The paragraph explains it already pretty well, imho.

Comment (dehora) [reply chapters/http-requests.adoc]: > I'd just link there. 

:+1:

**Comment by tkrop** [comment]
:+1:

File: chapters/http-requests.adoc (hunk 7 above)
Comment (ddufourboivin-zalando) [reply chapters/http-requests.adoc]: All for it. But I just noticed that [that rule](https://opensource.zalando.com/restful-api-guidelines/#_must_document_caching_if_supported) does not have a number. Is that normal?

Comment (ddufourboivin-zalando) [reply chapters/http-requests.adoc]: OK found the problem. There is no `#`. I am fixing this in this PR.

```adoc
[217]
== {MUST} Document Caching, if Supported
```

should be

```adoc
[#217]
== {MUST} Document Caching, if Supported
```

**Comment by dehora** [comment]
:+1:

**Comment by tkrop** [comment]
:+1:

**Pull request merged**